foxpuppet/windows/browser/notifications/addons.py

```diff
@@ -59,14 +59,57 @@ def close(self):
 class AddOnInstallRestart(BaseNotification):
     """Add-on install restart notification."""
 
+    def restart_now(self):
+        """Restart Firefox immediately to complete the add-on installation."""
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            self.find_primary_button().click()
+
+    def restart_later(self):
+        """Defer Firefox restart for later."""
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            self.find_secondary_button().click()
+
 
 class AddOnInstallFailed(BaseNotification):
     """Add-on install failed notification."""
 
+    @property
+    def error_message(self):
+        """Provide access to the error message.
+
+        Returns:
+            str: The error message explaining why the installation failed.
+        """
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            return self.find_description().text
+
+    def close(self):
+        """Close the failed installation notification."""
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            self.find_primary_button().click()
+
 
 class AddOnProgress(BaseNotification):
     """Add-on progress notification."""
 
+    @property
+    def is_downloading(self):
+        """Check if the add-on is currently downloading.
+
+        Returns:
+            bool: True if the download and verification is in progress.
+        """
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            return "Downloading and verifying add-on…" in self.find_description().text
+
+    def wait_until_complete(self, timeout=None):
+        """Wait until the progress notification disappears, indicating completion.
+
+        Args:
+            timeout (int, optional): Maximum time to wait in seconds.
+        """
+        self.window.wait_for_notification(None, timeout=timeout)
```

Comment:
Should this return something when it has been completed? Even returning the `self.window.wait_for_notification(None, timeout=timeout)` seems fine, as that returns the `notification` itself.

Reply:
Thanks for pointing that out. I think this could be potentially confusing if there is no return. Actually, from what I have observed of the Firefox notification for progress, the notification disappears once the download is completed. So I think it will be more accurate to return a boolean that shows that the notification is no longer active or visible.

```python

 return self.window.wait_for_notification(None, timeout=timeout) is None
 
 ```

Reply:
Well I notice that your call to `wait_for_notification` uses `None`, I think we would want to wait for this `AddOnProgress` notification right? I know that you can wait for any notification with that method, but it wouldn't seems right to have that be within this class. So do we need this method at all? When a download is completed there should be only a few different results from that, it seems to make sense to check for those. I think if you write a test for this it will help clarify what you had in mind.

Reply:
Actually, when a download completes, no result is usually displayed that shows the download has completed. When a download completes, the notification will no longer be visible. Usually, it is followed by the Add-on install complete notification or Blocked notification if it potentially suspicious extension.